### PR TITLE
Change `a` normalization scale to be more robust

### LINKED ISSUE #1953: Why do I get a QH when I optimize for a QA?
Here's a minimally working script with the latest version of DESC, using which I am trying to get a simple QA equilibrium.

I start with @f0uriest 's generalized parametrization and optimize for QA and some shape-related quantities. It takes about 10 min on a 40G A100 GPU.
 
Even though the maximum Quasisymmetry error goes from  1e-1 -> 1e-14 the final quasisymmetry clearly close to a helical symmetry even though I specify helicity = (1, 0). I have tried both QuasisymmetryTwoTerm and QuasisymmetryBoozer.

Is there a bug in the Quasisymmetry classes? Am I missing something?

```python3

#!/usr/bin/env python3

from desc import set_device

set_device("gpu")

import jax

import numpy as np
from desc.backend import jnp, jax
from desc.geometry import FourierRZToroidalSurface
from desc.equilibrium import EquilibriaFamily, Equilibrium
from desc.grid import LinearGrid, Grid
from desc.objectives import (
    ForceBalance,
    ObjectiveFunction,
    MagneticWell,
    FixBoundaryR,
    FixBoundaryZ,
    FixCurrent,
    FixPressure,
    FixPsi,
    AspectRatio,
    FixPressure,
    Elongation,
    RotationalTransform,
    BScaleLength,
    QuasisymmetryTwoTerm,
    QuasisymmetryBoozer,
)
from desc.optimize import Optimizer
from desc.profiles import PowerSeriesProfile, SplineProfile
from desc.magnetic_fields import OmnigenousField
from desc.vmec import VMECIO
from desc.plotting import *

from desc.integrals import Bounce2D
from desc.compat import rescale, flip_theta

from matplotlib import pyplot as plt
import os
import sys


NFP = 1
aspect_ratio = 7.0
twist = 1.0
torsion = 0.0
elongation = 3.0
triangularity = -0.5
squareness = 0.1
eccentricity = 0.0


mirror_ratio = 0.2
well_weight = 2.0

# setting these values manually
helicity = (1, 0)
L = int(10)
M = int(10)
N = int(10)


# dependent variables
M_grid = int(np.ceil(2.5 * M))
N_grid = int(np.ceil(2.5 * N))

R0 = 1.0
minor_radius = R0 / aspect_ratio
flux = np.pi * minor_radius**2

init_fname = f'NFP{NFP}_A{aspect_ratio:+.1f}_twist{twist:+.1f}_tor{torsion:+.1f}_elong{elongation:+.1f}_tri{triangularity:+.1f}_square{squareness:+.1f}_ecc{eccentricity:+.1f}'
fname = init_fname

p_l = [0, 0., 0., 0., -0., 0.]

Ntheta = 100
Nphi = 200

theta = np.linspace(0, 2*np.pi, Ntheta)
phi = np.linspace(0, 2*np.pi, Nphi)

theta_2D, phi_2D = np.meshgrid(theta, phi, indexing="ij")

surf = FourierRZToroidalSurface.from_shape_parameters(major_radius=R0, aspect_ratio=aspect_ratio, elongation=elongation, triangularity=triangularity, squareness=squareness, eccentricity=eccentricity, torsion=torsion,  twist=twist, NFP=NFP)
flux = np.pi * minor_radius ** 2

eq = Equilibrium(Psi=flux, NFP=NFP, L=L, M=M, N=N, sym=True, surface=surf)
eq, _ = eq.solve(objective="force", ftol=1e-4, xtol=1e-6, gtol=1e-6, maxiter=120)

eq.save("eq_" + init_fname + "_vacuum_init_QS.h5")

print(eq.compute("<beta>_vol")["<beta>_vol"])  

rho = np.linspace(0, 1, L+1)
grid = LinearGrid(L=L, M = M_grid, N=N_grid, NFP=eq.NFP, sym=True)
data_keys = ["iota", "D_Mercier", "magnetic well"]
data = eq.compute(data_keys, grid=grid)

print("D_Mercier=", grid.compress(data["D_Mercier"]));
print("magnetic well=", grid.compress(data["magnetic well"]));
print("iota", grid.compress(data["iota"]))

fname = init_fname

surfaces_omni = [0.25, 0.4, 0.6, 0.85] 
QS_weights = [20, 40.0, 60, 80, 100, 150]
mirror_weights = [4e2, 7e2, 1e3, 2e3]
scale_weights = [4e2, 4e2, 8e2, 1e3]


start_idx = 1
gap = 1

staging_array = [1, 2, 3, 4]

for k in staging_array:
   
    print("\n---------------------------------------")
    print(f"Optimizing boundary modes M, N <= {k}")
    print("---------------------------------------")
    

    try:
        QS_weight = QS_weights[int(k-1)]
    except:
        QS_weight = QS_weights[-1]

    try:
        mirror_weight = mirror_weights[int(k-1)]
    except:
        mirror_weight = mirror_weights[-1]
   
    try:
        scale_weight = scale_weights[int(k-1)]
    except:
        scale_weight = scale_weights[-1]


    Elongation_grid = LinearGrid(M = 2*int(M), N = 2*int(N), rho=np.array([1.0]), NFP=eq.NFP, sym=False, axis=False)
    iota_grid = LinearGrid(M = 2*int(M), N = 2*int(N), rho=np.array([0.05]), NFP=eq.NFP, sym=True, axis=False)
    QS_grid = LinearGrid(
        M=M_grid, N=eq.N_grid, NFP=eq.NFP, rho=surfaces_omni, sym=False
    )

    modes_R = np.vstack(
        (
            [0, 0, 0],
            eq.surface.R_basis.modes[
                np.max(np.abs(eq.surface.R_basis.modes), 1) > k, :
            ],
        )
    )
    modes_Z = eq.surface.Z_basis.modes[
        np.max(np.abs(eq.surface.Z_basis.modes), 1) > k, :
    ]
    
    objective = ObjectiveFunction((ForceBalance(eq=eq, weight = 1e4),
                                   AspectRatio(eq=eq, bounds=(aspect_ratio-2.5, aspect_ratio+1.0), weight=1e4),
                                   Elongation(eq=eq, bounds=(3.0, 8.), grid=Elongation_grid, weight=5e2),
                                   RotationalTransform(eq=eq, bounds=(0.3, np.inf), grid=iota_grid, weight=5e3),
                                   BScaleLength(eq=eq, bounds=(0.2 + 0.02*(10-eq.NFP) - 0.004*aspect_ratio, np.inf), grid=Elongation_grid,\
                                                 normalize=False, normalize_target=False,  weight=scale_weight),
                                   QuasisymmetryTwoTerm(eq=eq, helicity=(1, 0), grid=QS_grid, weight=QS_weight)))
                                   #QuasisymmetryBoozer(eq=eq, helicity=(1, 0), grid=QS_grid, weight=QS_weight)))
    constraints = (
            ForceBalance(eq=eq),
            FixBoundaryR(eq=eq, modes=modes_R),
            FixBoundaryZ(eq=eq, modes=modes_Z),
            FixPressure(eq=eq),
            FixCurrent(eq=eq),
            FixPsi(eq=eq),
            )
    optimizer = Optimizer("proximal-lsq-exact")
    
    eq, _ = eq.optimize(
                objective,
                constraints,
                ftol=5e-4,
                xtol=1e-9,
                gtol=1e-9,
                maxiter=48,
                verbose=3
                )

    
    eq.save("eq_" + fname + f"_vacuum_{k:02d}_QS.h5")

    jax.clear_caches()



L0 = 10
rho = np.linspace(0, 1, L0+1)
grid = LinearGrid(L=L0, M = M_grid, N=N_grid, NFP=eq.NFP, sym=True)
data_keys = ["iota"]
iota = eq.compute(data_keys, grid=grid)["iota"]

eq.save("eq_" + fname + f"_vacuum_final2_QS.h5")

print("Final equilibrium nestedness = ", eq.is_nested())

fig, ax = plot_section(eq, "|F|", norm_F=True, log=True); 
plt.savefig(f"normF_vacuum_QS.png", dpi=400)  
plt.close()

eq0 = Equilibrium.load("eq_" + init_fname + "_vacuum_init_QS.h5")

fig, ax = plot_comparison(eqs=[eq0, eq]) 
plt.savefig(f"Xsection_vacuum_QS.png", dpi=400)  
plt.close()

fig, ax = plot_boozer_surface(eq, rho=1)
plt.savefig(f"Boozer_eq_vacuum_QS.png", dpi=400)  
plt.close()

```

### LINKED ISSUE #1954: `"a"` normalization can give bad results in some cases
In this reasonable-looking boundary (obtained from the step-one of #1953 script), 

[input_eq_with_bad_a_normalization.txt](https://github.com/user-attachments/files/22686017/input_eq_with_bad_a_normalization.txt)

`desc.objectives.normalization.compute_scaling_factors(eq)` yields insane values for some downstream normalizations bc its estimate for `a` is so off

```python
from desc.objectives.normalization import compute_scaling_factors
from desc.equilibrium import Equilibrium
eq = Equilibrium.from_input_file("input_eq_with_bad_a_normalization.txt")
compute_scaling_factors(eq)
```

```python
{'R0': Array(1., dtype=float64),
 'a': np.float64(0.0014579891339794435),
 'Psi': 0.0641141358,
 'A': np.float64(6.678185023680789e-06),
 'V': Array(4.1960274e-05, dtype=float64),
 'B': np.float64(12000.666268726422),
 'I': np.float64(60003331351.554504),
 'W': Array(2.40441358e+09, dtype=float64),
 'J': np.float64(6549998600137.333),
 'F': np.float64(3.930217363043669e+16),
 'f': Array(1.64912998e+12, dtype=float64),
 'p': np.float64(57302142094950.1),
 'W_p': Array(1.20220679e+09, dtype=float64)}
```

At this point, do we just populate it with `eq.compute("a")` in `compute_scaling_factors`? this is not the first time we've had to deal with this #821

for ref this is the eq: finite radius a def not ~1e-3m, and |B| that is nowhere near 1e4 T

Image

Image

### PR DESCRIPTION
**Pull request by dpanici**
Resolves #1954
Resolves #1953

### CODE REVIEW

**Review by rahulgaur104 — DISMISSED** [review]

File: desc/objectives/normalization.py [hunk 1]
```diff
@@ -15,33 +15,34 @@ def compute_scaling_factors(thing):
     scales = {}
 
     def get_lowest_mode(basis, coeffs):
-        """Return lowest order coefficient (excluding m=0 modes)."""
+        """Return average (over n) of M=+/1 modes."""
```
Comment (f0uriest) [inline desc/objectives/normalization.py]: how is this the average? isn't it the max?

**Review by f0uriest — APPROVED** [review]

File: desc/objectives/normalization.py (hunk 1 above)
Comment (dpanici) [reply desc/objectives/normalization.py]: mistype in the docstring, my bad

File: desc/objectives/normalization.py [hunk 2]
```diff
@@ -14,34 +14,34 @@ def compute_scaling_factors(thing):
...
+    def get_max_mode(basis, coeffs):
+        """Return max boundary mode that is not M=0 N=0."""
+        if basis.N > 0:
+            inds_not_00 = np.where(
+                np.logical_or(
+                    np.abs(basis.modes[:, 1]) > 0, np.abs(basis.modes[:, 2]) > 0
+                )
+            )[0]
+        else:
+            inds_not_00 = np.where(np.abs(basis.modes[:, 1]) > 0)[0]
```
Comment (ddudt) [inline desc/objectives/normalization.py]: I don't think we need this "else" condition, the above general case should also work for N=0.

File: tests/test_examples.py [hunk 3]
```diff
@@ -807,7 +807,10 @@ def test_NAE_QSC_solve_near_axis_asym():
 
         # check |B| on axis
         np.testing.assert_allclose(
-            data_nae["|B|"], np.ones(np.size(phi)) * qsc.B0, atol=1e-4, err_msg=string
+            data_nae["|B|"],
+            np.ones(np.size(phi)) * qsc.B0,
+            atol=2e-4,
+            err_msg=string,
```
Comment (ddudt) [inline tests/test_examples.py]: remove the trailing comma so this stays on one line
